lib/galaxy/managers/datasets.py

```diff
@@ -684,6 +683,20 @@ def deserialize_metadatum(self, dataset_assoc, key, val, **context):
         # ...?
         return unwrapped_val
 
+    def deserialize_datatype(self, item, key, val, **context):
+        if not item.datatype.allow_datatype_change:
+            raise exceptions.RequestParameterInvalidException("The current datatype does not allow datatype changes.")
+        if not self.app.datatypes_registry.get_datatype_by_extension(val):
+            raise exceptions.RequestParameterInvalidException("The target datatype does not exist.")
+        if not self.app.datatypes_registry.get_datatype_by_extension(val).allow_datatype_change:
+            raise exceptions.RequestParameterInvalidException("The target datatype does not allow datatype changes.")
+        if not DatasetAssociationManager(self.app).ok_to_edit_metadata(item.dataset_id):
+            raise Exception("Dataset metadata could not be updated, probably because it is not in an 'ok' state.")
```

Comment:
```suggestion
            raise Exception("Dataset metadata could not be updated, probably because it is used as input or output.")
```

Reply:
`used as input or output` is not clear for me - I only encountered this error so far when a dataset is uploading (like in the test) or a previous datatype change is not complete

Reply:
I just edited the message, `used as input or output of a running job.` which is what you're describing, right ?